### PR TITLE
Fesystem indexing

### PR DESCRIPTION
**Pull request by drwells**
There's a curious comment in `fe_system.cc`:
```cpp
1334        // TODO: Introduce the needed table and loop only over base element
1335        // shape functions. This here is not efficient at all AND very bad style
```
I noticed that some of my own programs were spending about 12% of their total runtime in this function which performs a lot of index calculations over and over again. I implemented index caching for both primitive and non-primitive elements - this sped things up by about 50%.

### CODE REVIEW

File: include/deal.II/fe/fe_system.h [hunk 1]
```diff
@@ -1056,6 +1049,39 @@ class FESystem : public FiniteElement<dim, spacedim>
...
+    /**
+     * Index into the base element's
+     * FEValuesImplementation::FiniteElementRelatedData arrays for the current
+     * base element shape function.
+     */
+    unsigned int in_index;
+
+    /**
+     * Index into the present element's
+     * FEValuesImplementation::FiniteElementRelatedData arrays for the current
```
Comment (tjhei) [inline include/deal.II/fe/fe_system.h:1079]: this comment is identical to above, copy&paste mistake?

File: include/deal.II/fe/fe_system.h [hunk 2]
```diff
@@ -14,32 +14,25 @@
 // ---------------------------------------------------------------------
 
 #ifndef dealii_fe_system_h
-#  define dealii_fe_system_h
+#define dealii_fe_system_h
```
Comment (tjhei) [inline include/deal.II/fe/fe_system.h:17]: why did this change?

File: source/fe/fe_system.cc [hunk 3]
```diff
@@ -42,8 +42,185 @@ namespace
...
+    const unsigned int n_components = fe.element_multiplicity(base_no);
+    const unsigned int n_dofs_per_cell =
+      fe.base_element(base_no).n_dofs_per_cell();
+    for (unsigned int component = 0; component < n_components; ++component)
+      for (unsigned int b = 0; b < n_dofs_per_cell; ++b)
+        {
+          const unsigned int out_index = base_to_system_table[component][b];
+
+          if (base_flags & update_values)
+            for (unsigned int q = 0; q < n_q_points; ++q)
```
Comment (tjhei) [inline source/fe/fe_system.cc:152]: Would it be faster to swap the ``for`` with the ``if`` and combine the loops?

File: tests/fe/fe_system_values.cc [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    FESystem<2> fe(FE_RaviartThomas<2>(0),
+                   1,
+                   FE_Q<2>(2),
+                   2,
+                   FE_RaviartThomas<2>(2),
+                   3,
+                   FE_Nothing<2>(),
+                   4,
+                   FE_DGQ<2>(0),
+                   1);
```
Comment (tjhei) [inline tests/fe/fe_system_values.cc:90]: what a beautiful FE! :-)

**Review by tjhei — APPROVED** [review]
looks good. I just have some minor comments.

File: include/deal.II/fe/fe_system.h (hunk 2 above)
Comment (drwells) [reply include/deal.II/fe/fe_system.h:17]: We don't 'indent' preprocessor directives in most of the other files (like `fe_q.h`) so I changed this to match that style.

File: include/deal.II/fe/fe_system.h (hunk 1 above)
Comment (drwells) [reply include/deal.II/fe/fe_system.h:1079]: It isn't identical: this is for the present element, the first is for the base element (one word is different).

File: source/fe/fe_system.cc (hunk 3 above)
Comment (drwells) [reply source/fe/fe_system.cc:152]: Maybe. I'll try it out.

Comment (drwells) [reply source/fe/fe_system.cc:152]: I timed moving switching the flag check and the quadrature point loop and also using `std::copy()` or changing other loop orderings. None of them were faster than the version in the original patch (in fact, they were all pretty close).

File: source/fe/fe_system.cc [hunk 5]
```diff
@@ -42,8 +42,185 @@ namespace
...
+          const unsigned int base_component =
+            fe.system_to_base_index(system_index).first.second;
+          const unsigned int base_index =
+            fe.system_to_base_index(system_index).second;
+          Assert(base_index < fe.base_element(base_no).n_dofs_per_cell(),
+                 ExcInternalError());
+
+          unsigned int out_index = 0;
+          for (unsigned int i = 0; i < system_index; ++i)
+            out_index += fe.n_nonzero_components(i);
```
Comment (kronbichler) [inline source/fe/fe_system.cc:79]: While you only pay this cost at setup, I still think we should avoid this second loop (making this a quadratic complexity algorithm). I think you could avoid this by defining `out_index` before line 64 and incrementing it by `fe.n_nonzero_components(system_index)` after the use in line 80.

File: source/fe/fe_system.cc (hunk 3 above)
Comment (kronbichler) [reply source/fe/fe_system.cc:152]: I think this is the most functions in the `FEValues` context work this way, and I think the code is pretty clean this way. One could reason about whether it is actually necessary to copy around all those arrays, but that is a fundamental design choice of how `FEValues` interacts with `FESystem`. Even though I personally find that inefficient, that ship has sailed, and we have other concepts for higher order shape functions where the cost is significant (matrix-free module).

File: source/fe/fe_system.cc (hunk 5 above)
Comment (drwells) [reply source/fe/fe_system.cc]: Fixed!

File: source/fe/fe_system.cc (hunk 3 above)
Comment (tjhei) [reply source/fe/fe_system.cc:154]: Thank you for trying it out!

File: include/deal.II/fe/fe_system.h (hunk 2 above)
Comment (tjhei) [reply include/deal.II/fe/fe_system.h:17]: I thought clang-format does this indentation? 😕

**Review by kronbichler — APPROVED** [review]
Nice, thank you for the update.

**Review by masterleinad — APPROVED** [review]

File: include/deal.II/fe/fe_system.h (hunk 2 above)
Comment (drwells) [reply include/deal.II/fe/fe_system.h:17]: I don't fully understand its behavior, but it definitely won't modify the original indentation scheme used in this file for `#include`s at the top.

**Pull request merged**